### DISCUSSION TITLE
Different http port?

### DISCUSSION THREAD

**Discussion by richelo** [ricklamers/gpt-code-ui#10] [Q&A]
Hi,

Love what I am seeing so far! GREAT JOB!

The main system I want to run this on is already using port 8080. How can I start it and tell it to use a different port? Preferably without editing the source code.

**Comment by ricklamers** [comment]
Thank you @richelo. `WEB_PORT` and `API_PORT` environment variables are now available to configure the ports used by the two Flask apps. Note, `API_PORT` doesn't have to be exposed, just bound locally (in case you're exposing the app someway, be careful! you don't want to expose this app to the public network).

**Comment by richelo** [comment]
Thank you so much for implementing this!

It does not seem to work though. I did upgrade to the latest version. No matter what port I try, I get the same error every time. I have confirmed that the ports I am trying are not currently in use on the system.

gptcode
[2023-05-22 08:59:46,043] ERROR [root:30] Error running the kernel_program:
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/dist-packages/gpt_code_ui/main.py", line 28, in run_kernel_program
    asyncio.run(kernel_program_main())
  File "/usr/lib/python3.9/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/lib/python3.9/asyncio/base_events.py", line 642, in run_until_complete
    return future.result()
  File "/usr/local/lib/python3.9/dist-packages/gpt_code_ui/kernel_program/main.py", line 146, in main
    await start_snakemq()
  File "/usr/local/lib/python3.9/dist-packages/gpt_code_ui/kernel_program/main.py", line 68, in start_snakemq
    messaging, link = utils.init_snakemq(config.IDENT_MAIN)
  File "/usr/local/lib/python3.9/dist-packages/gpt_code_ui/kernel_program/utils.py", line 24, in init_snakemq
    link.add_listener(("localhost", config.SNAKEMQ_PORT))
  File "/usr/local/lib/python3.9/dist-packages/snakemq/link.py", line 343, in add_listener
    listen_sock.listen(address)
  File "/usr/local/lib/python3.9/dist-packages/snakemq/link.py", line 93, in listen
    self.sock.bind(address)
OSError: [Errno 98] Address already in use

█▀▀ █▀█ ▀█▀ ▄▄ █▀▀ █▀█ █▀▄ █▀▀
█▄█ █▀▀ ░█░ ░░ █▄▄ █▄█ █▄▀ ██▄

> Open GPT-Code UI in your browser http://localhost:8090

**Comment by ricklamers** [comment]
https://github.com/ricklamers/gpt-code-ui/blob/8fa48dd2c59e192a2d595be12597f1df56c73a96/gpt_code_ui/kernel_program/config.py#L7

Might be the 8765 port used by SnakeMQ? If you had any prior failed starts make sure to kill all of those running processes (you could use a quick hack like `killall python`)

**Comment by ricklamers** [comment]
Made this port configurable with the `SNAKEMQ_PORT` environment variable now too.

**Comment by richelo** [comment]
Update to the latest version.

Set the following in .env:
SNAKEMQ_PORT=8766

Checked that the port was not in use:
sudo lsof -i:8766
Returns nothing, so not in use.

Then ...

gptcode
[2023-05-22 10:29:32,934] ERROR [root:30] Error running the kernel_program:
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/dist-packages/gpt_code_ui/main.py", line 28, in run_kernel_program
    asyncio.run(kernel_program_main())
  File "/usr/lib/python3.9/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/lib/python3.9/asyncio/base_events.py", line 642, in run_until_complete
    return future.result()
  File "/usr/local/lib/python3.9/dist-packages/gpt_code_ui/kernel_program/main.py", line 146, in main
    await start_snakemq()
  File "/usr/local/lib/python3.9/dist-packages/gpt_code_ui/kernel_program/main.py", line 68, in start_snakemq
    messaging, link = utils.init_snakemq(config.IDENT_MAIN)
  File "/usr/local/lib/python3.9/dist-packages/gpt_code_ui/kernel_program/utils.py", line 24, in init_snakemq
    link.add_listener(("localhost", config.SNAKEMQ_PORT))
  File "/usr/local/lib/python3.9/dist-packages/snakemq/link.py", line 343, in add_listener
    listen_sock.listen(address)
  File "/usr/local/lib/python3.9/dist-packages/snakemq/link.py", line 93, in listen
    self.sock.bind(address)
OSError: [Errno 98] Address already in use

█▀▀ █▀█ ▀█▀ ▄▄ █▀▀ █▀█ █▀▄ █▀▀
█▄█ █▀▀ ░█░ ░░ █▄▄ █▄█ █▄▀ ██▄

> Open GPT-Code UI in your browser http://localhost:8090

Looks like it is not using the SNAKEMQ_PORT.

**Reply by ricklamers** [reply]
Ah, I see the confusion. The `.env` is not sourced for the `*_PORT` environment variables in all processes. I assumed people would source the `.env` in their shell or set the environment variables differently. Let me add the required `load_dotenv('.env')`.

**Reply by ricklamers** [reply]
`v0.42.26` should handle this through the `.env` file now too.

**Comment by richelo** [comment]
Apologies for making more work for you!

Thank you, though, for being so accommodating! MUCH appreciated!

Maybe as a thought ... Make a docker-compose version where the ports can then be mapped in the docker-compose.yml file. Then there is no need for doing it in env, etc.

**Reply by ricklamers** [reply]
I wanted to make this a Python only package such that others can choose to run this in containers as they see fit.

**Comment by richelo** [comment]
I actually exported both ports in the shell, and now see no errors when running gptcode, but when I try and access it via the browser it does not work.

The browser is not on the same machine, so accessing it via 192.168.1.x:8090, but no joy. When I load gptcode local on my Mac, and then access via localhost, that works.

I have several docker containers running on the Linux machine and access it with no problem via IP.

**Reply by ricklamers** [reply]
Can you try with `0.42.26`? I think something may have gone wrong in the bundling process (some `VITE_` dev vars leaked in)

**Reply by richelo** [reply]
Nope, still getting a connection refused on 0.42.26.

**Reply by richelo** [reply]
Could you try accessing it from a different machine to what it is running on. Just want to know if it's the software, or my setup. Probably my issue, but just want to be sure. ;-)

**Reply by richelo** [reply]
Also ... Sorry for the totally newb question ... How do I get to the logs to see if I see any error there?

**Reply by richelo** [reply]
FYI ... Installed on a RPI that has nothing else running. Using all standard ports.

Still get the same issue ... Cannot access it via ip:8080

**Reply by ricklamers** [reply]
I had to broaden host network binding in Flask. Please try `0.42.28`*.

**Reply by ricklamers** [reply]
Error logs should be in `app.log` in the working directory in which you run `gptcode`.

**Reply by richelo** [reply]
YES!!! 0.42.28 works perfectly on ip:port!

THANK YOU!

**Reply by ricklamers** [reply]
Glad it worked ;-) Spread the word 🤙
